chunky/src/java/se/llbit/chunky/world/JavaChunk.java

```diff
@@ -0,0 +1,377 @@
+package se.llbit.chunky.world;
+
+import it.unimi.dsi.fastutil.ints.IntIntImmutablePair;
+import se.llbit.chunky.block.legacy.LegacyBlocks;
+import se.llbit.chunky.chunk.BlockPalette;
+import se.llbit.chunky.chunk.ChunkData;
+import se.llbit.chunky.chunk.ChunkLoadingException;
+import se.llbit.chunky.chunk.EmptyChunkData;
+import se.llbit.chunky.chunk.biome.BiomeDataFactory;
+import se.llbit.chunky.map.BiomeLayer;
+import se.llbit.chunky.map.IconLayer;
+import se.llbit.chunky.map.SurfaceLayer;
+import se.llbit.chunky.world.biome.ArrayBiomePalette;
+import se.llbit.chunky.world.biome.BiomePalette;
+import se.llbit.chunky.world.region.MCRegion;
+import se.llbit.log.Log;
+import se.llbit.math.QuickMath;
+import se.llbit.nbt.CompoundTag;
+import se.llbit.nbt.ListTag;
+import se.llbit.nbt.SpecificTag;
+import se.llbit.nbt.Tag;
+import se.llbit.util.BitBuffer;
+import se.llbit.util.Mutable;
+import se.llbit.util.annotation.NotNull;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static se.llbit.util.NbtUtil.getTagFromNames;
+import static se.llbit.util.NbtUtil.tagFromMap;
+
+public class JavaChunk extends Chunk {
+  public static final String DATAVERSION = ".DataVersion";
+  public static final String LEVEL_HEIGHTMAP = ".Level.HeightMap";
+  public static final String LEVEL_SECTIONS = ".Level.Sections";
+  public static final String LEVEL_BIOMES = ".Level.Biomes";
+  public static final String LEVEL_ENTITIES = ".Level.Entities";
+  public static final String ENTITIES_POST_20W45A = ".Entities";
+  public static final String LEVEL_TILEENTITIES = ".Level.TileEntities";
+  public static final String BLOCK_ENTITIES_POST_21W43A = ".block_entities";
+
+  public static final String SECTIONS_POST_21W39A = ".sections";
+  public static final String BIOMES_POST_21W39A = ".biomes";
+
+  public static final int SECTION_BYTES = X_MAX * SECTION_Y_MAX * Z_MAX;
+  public static final int SECTION_HALF_NIBBLES = SECTION_BYTES / 2;
+  private static final int CHUNK_BYTES = X_MAX * Y_MAX * Z_MAX;
+
+  public static final int DATAVERSION_20W17A = 2529;
+  public static final int DATAVERSION_20W45A = 2681; // entities moved into separate region files
+
+  public JavaChunk(ChunkPosition pos, JavaDimension dimension) {
+    super(pos, dimension);
+  }
+
+  /**
+   * @param request fresh request set
+   * @return loaded data, or null if something went wrong
+   */
+  private Map<String, Tag> getChunkTags(Set<String> request) throws ChunkLoadingException {
```

Comment:
Off topic: I never really understood why we had this request/response interface for getting chunk tags.

Reply:
yeah me neither, great flexibility but it's unnecessary in our usecase